src/rgw/rgw_common.cc

```diff
@@ -283,10 +492,29 @@ void req_info::init_meta_info(bool *found_bad_meta)
 
 std::ostream& operator<<(std::ostream& oss, const rgw_err &err)
 {
-  oss << "rgw_err(http_ret=" << err.http_ret << ", s3='" << err.s3_code << "') ";
+  oss << "rgw_err(http_ret=" << err.http_ret << ", s3='" << err.get_err_code() << "') ";
   return oss;
 }
 
+void rgw_err::dump() const
+{
+  const string& err_code = get_err_code();
+  const string& err_message = get_err_message();
+  if (s->format != RGW_FORMAT_HTML)
+    s->formatter->open_object_section("Error");
+  if (!err_code.empty())
+    s->formatter->dump_string("Code", err_code);
+  if (!err_message.empty())
+    s->formatter->dump_string("Message", err_message);
+  if (!s->bucket_name.empty())	// TODO: connect to expose_bucket
+    s->formatter->dump_string("BucketName", s->bucket_name);
+  if (!s->trans_id.empty())	// TODO: connect to expose_bucket or another toggle
+    s->formatter->dump_string("RequestId", s->trans_id);
+  s->formatter->dump_string("HostId", s->host_id);
+  if (s->format != RGW_FORMAT_HTML)
+    s->formatter->close_section();
+}
+
```

Comment:
i *think* `rgw_err::dump()` is the only part of `rgw_err` that depends on having this `req_state *s` pointer. can we just add that as an argument to `dump()`, so that we don't need to store it inside of `rgw_err` and its variant types?